### ISSUE TITLE
data preprocess tool using hdf5 or tfrecord

### ISSUE THREAD

**Issue by tongjinle123** [PyTorchLightning/pytorch-lightning#1231]
## 🚀 Feature
A subpackage or tool using `hdf5` or `tfrecord` to preprocess data into one single file.
### Motivation
In some field like asr or cv, it is not very novel to just use pytorch dataloader because it may cause speed loss in online data process like making fbank `feature(asr)` or some `transforms(cv)`. And hdf5 or tfrecord can be a good choice to avoid IO bottleneck and cpu bottleneck.  And I think it could be much helpful that our project can have a sub package or tool to do that---either write and read. And there is a texar-pytorch have made such function see: 
https://texar-pytorch.readthedocs.io/en/latest/code/data.html#recorddata

also, dataloder utils should be adapted to this because this may need to use iterable dataset plus using `num_workers > 0` in dataloader and the missing of the length of the dataset can be a problem for the training process. 

### Pitch

the link above can be an example but there still a need for writing and loading var length processed `feature(tensor dim like [1, sequence_length, feature_dim])` in using `hdf5(this can be a little complex)`

I tried to write a little tool for this intention https://github.com/tongjinle123/tfrecord_builder
but when I was using it in our project some months ago, I found it hard to use it directly because the iterable dataset is hard to use.

also, there are some awesome tools for this intention like:  https://github.com/vahidk/tfrecord
### Additional context
It can be much helpful that our project can take this into consideration and please forgive my bad English   : )
I hope I have fully expressed my idea.

**Comment by Borda** [comment]
I like the idea of extending `prepare_data` to prepared some hdf5 file which would be used in later reruns... any thoughts @PyTorchLightning/core-contributors

**Comment by tongjinle123** [comment]
plus: 
for using hdf5, there is a problem occurred when using num_worker>0 pytorch dataloader and here is the solution, hope that can be helpful

**Comment by Borda** [comment]
Any other suggestion for using different filesystem?

**Comment by tongjinle123** [comment]
when using hdf5, in my practice, it seems that there are two ways for storing var length feature.
1. encode the feature with np.byte() and decode in dataset when loading. this may be a little bit slower process and I do not like it. (we also have to store a decode policy and decoding makes the process dirty)
2. initialize a [num_data, max_length, feature_dim] array in hdf5 and then store features (or initialize a bunch of array).
and we should not simply process all the features and fill in the hdf5, we should make the process online, for memory related reason.

talking about other filesystem, I have not find any...

**Comment by williamFalcon** [comment]
it would be cool to add tfrecord and hdf5 support to lightning.

Maybe we can do it this way:
1. enable train_dataloader to also return hdf5 and tfrecord.
2. internally, lightning detects these and wraps in dataloaders?
3. the dataloaders are classes we write to interface with these other data types?

**Comment by Skylion007** [comment]
Pytorch XLA has [first party support](https://github.com/pytorch/xla/blob/master/torch_xla/utils/tf_record_reader.py) for reading tf records now. We should just wrap that.
